lib/puppet/face/config.rb

```diff
@@ -106,6 +106,19 @@
     EOT
 
     when_invoked do |name, value, options|
+      if name == 'environment' && options[:section] == 'main'
+        Puppet.warning _(<<-EOM).chomp
+The environment should be set in either the `[user]`, `[agent]`, or `[master]`
+section. Variables set in the `[agent]` section are used when running
+`puppet agent`. Variables set in the `[user]` section are used when running
+various other puppet subcommands, like `puppet apply` and `puppet module`. These
```

Comment:
Super small nitpick, but I'd use a semicolon rather than period to make it clear that applies only to the previous sentence.

Reply:
Or the `, and` linking I originally used.

Reply:
semicolon added